### ISSUE TITLE
Identical non-primitive values don't throw type errors when mixed incorrectly

### ISSUE THREAD

**Issue by fusepilot** [Microsoft/TypeScript#12899]
**TypeScript Version:**  2.1.4

**Code**

```ts
type Foo = {
	name: string
}

type Bar = {
	name: string
}

const bar: Bar = {
	name: 'barA',
}

const fooArray: Foo[] = []
fooArray.push(bar) // no type error

function test(bar: Bar, action): Foo {
  return bar // no type error
}
```

**Expected behavior:**
Typescript should warn me that I'm mixing Foo and Bar together even though they are structurally identical and compile fine.

**Actual behavior:**
Typescript should know and warn me that I probably don't want to conflate Foo and Bar as this could lead to confusion if one of the types were to ever change.

I experienced this problem while prototyping the structure of my application. I hadn't fully fleshed the types out yet and was confused why I wasn't getting type any errors when I accidentally tried to add one type to an array of another type.

**Comment by mhegazy** [comment]
TypeScript has a structural (and not nominal) type system. please see https://github.com/Microsoft/TypeScript/wiki/FAQ#what-is-structural-typing for more details.
